src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -8,393 +8,89 @@
+	
+	init_basis(num_features);
+	m_fitted = true;
 }
 
 SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
 {
-	if (!test_rfinited()) {
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
-
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
-
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
-
-	return res;
+	return (SGVector<float64_t>)apply_to_matrix((SGMatrix<float64_t>) vector);
 }
 
-SGMatrix<float64_t>
-RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
+SGMatrix<float64_t> RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
 {
-	// version for case dim_feature_space < dim_input space with direct transformation on feature matrix ??
-
 	auto num_vectors = matrix.num_cols;
 	auto num_features = matrix.num_rows;
 
-	io::info("get Feature matrix: {}x{}", num_vectors, num_features);
+	assert_fitted();
+	SG_DEBUG("Got Feature matrix: {}x{}", num_vectors, num_features);
 
-	if (num_features != cur_dim_input_space)
-	{
-		throw ShogunException(
-		    "float64_t * "
-		    "CRandomFourierGaussPreproc::apply_to_matrix("
-		    "SGMatrix<float64_t> matrix): matrix.num_rows != "
-		    "cur_dim_input_space is not allowed\n");
-	}
+	SGMatrix<float64_t> projection(m_dim_output, num_vectors);
 
-	SGMatrix<float64_t> res(cur_dim_feature_space, num_vectors);
+	linalg::matrix_prod(m_basis,matrix,projection);
+	linalg::add_vector(projection, m_offset, projection);
+	linalg::cos(projection, projection);
+	linalg::scale(projection, projection, std::sqrt(2.0 / m_dim_output));
 
-	auto val = std::sqrt(2.0 / cur_dim_feature_space);
+	return projection;	
+}
 
-	for (auto vec : range(num_vectors))
+SGMatrix<float64_t> RandomFourierGaussPreproc::sample_spectral_density(int32_t dim_input_space) const
+{
+	NormalDistribution<float64_t> normal_dist;
+	auto sampled_kernel = SGMatrix<float64_t>(m_dim_output, dim_input_space);
+	for (auto row : range(m_dim_output))
 	{
-		for (auto od : range(cur_dim_feature_space))
+		for (auto col : range(dim_input_space))
 		{
-			SGVector<float64_t> a(
-			    matrix.matrix + vec * num_features, cur_dim_input_space, false);
-			SGVector<float64_t> b(
-			    randomcoeff_multiplicative + od * cur_dim_input_space,
-			    cur_dim_input_space, false);
-			res(od, vec) =
-			    val * cos(randomcoeff_additive[od] + linalg::dot(a, b));
+			auto width = std::exp(m_log_width * 2.0) * 2.0;
```

Comment:
```suggestion
			const auto width = std::exp(m_log_width * 2.0) * 2.0;
```

Reply:
I am not sure the compiler will optimise this, but you might be better off computing this outside the loop?

Reply:
The compiler should optimize variable declaration inside loop.

Reply:
are you sure? because m_log_width could be changed in another thread, so how can the compiler assume this will always be the same?

Reply:
you do have a point. i guess putting the const width inside loop doesn't make too much sense either. Okay then

Reply:
btw I think the optimisations it could do here is reuse the memory allocation on the stack and then the CPU itself can see that you are recomputing the same all the time and put the result in the hot path. If m_log_width was const, then I think the compiler would compute the value only once.

Reply:
making data members as const would render the setter useless, wouldn't it? I've put width outside the loop, and I think that's for better.

Reply:
i am just telling you when there could be optimisations. You could imagine copying it to a const variable to test it out. But yes, you should put it outside the loop to be sure.

Reply:
I see, thanks. No more loops btw :)